crates/abd-clam/src/chaoda/cluster.rs

```diff
@@ -16,8 +16,18 @@ use serde::{
 
 use crate::{core::cluster::Children, utils, Cluster, Dataset, Instance, PartitionCriterion, UniBall};
 
+/// An extension of the `Cluster` trait to allow for anomaly detection.
+pub trait OddBall<U: Number>: Cluster<U> {
+    /// Return a 2-tuple of the properties of the cluster used for anomaly detection,
+    /// and the exponential moving averages of those properties.
+    fn ratios(&self) -> Vec<f32>;
+
+    /// Return the accumulated child-parent cardinality ratio.
+    fn accumulated_cp_car_ratio(&self) -> f32;
+}
+
 /// The ratios used for anomaly detection.
-pub type Ratios = [f64; 6];
+pub type Ratios = [f32; 6];
```

Comment:
`f32` has enough precision for our purposes and offers a significant speed bump for CHAODA.